### ISSUE TITLE
NOTICE:  bogus varno: 65001 (ruleutils.c:3229)

### ISSUE THREAD

**Issue by vincexiaoqiang** [greenplum-db/gpdb#1708]
Hi
   When i explain analyze sql, shows up 'bogus varno: 65001';

   here are steps: 
   1.    CREATE TABLE test(
              organ_no character varying(20), -- 机构编码
              store_no character varying(20), -- 店仓编码
              product_no character varying(20), -- 商品编码
              brand_no character varying(20), -- 品牌部编码
              inv_flag numeric(5,0), -- bi库存标志(0 正品; 1 次品)
              bill_date character varying(8)
        )
        WITH (APPENDONLY=true, COMPRESSLEVEL=5, COMPRESSTYPE=zlib, 
        OIDS=FALSE
       )
       DISTRIBUTED BY (organ_no, store_no, product_no)
       PARTITION BY RANGE(bill_date) 
          (
                 PARTITION p201212  END ('20130101'::character(8)), 
                 PARTITION p201301 START ('20130101'::character(8)) END ('20130201'::character(8))
         );

     2.   insert into test select 'AA','AA','AA','AA',1,'20130101' from generate_series(1,10000);

     3.  analyze test ;  (necessary)

     4.  set optimizer=on;

     5.  explain analyze
          select t2.organ_no
            from test t2
          group by t2.organ_no;

      6. NOTICE:  bogus varno: 65001 (ruleutils.c:3229)
          ********** 错误 **********

**Comment by vincexiaoqiang** [comment]
Additional instructions：
version: 4.3.99.00 build dev
commit id: 38efcba47b718244d6d7a91379fc4f94c5be2d10

**Comment by hlinnaka** [comment]
Works for me, on git master. Perhaps this was fixed some time after commit 38efcba?

**Comment by vraghavan78** [comment]
@hlinnaka I am getting the following:

```
vraghavan=# set optimizer=on;                                                                                                                                                                                                                                                                                                                                             SET
vraghavan=# explain select t2.organ_no from test t2 group by t2.organ_no; 
NOTICE:  bogus varno: 65001 (ruleutils.c:3228)
```

**Comment by hlinnaka** [comment]
On 02/08/2017 09:36 PM, Venkatesh Raghavan wrote:
> @hlinnaka I am getting the following:
>
> ```
> vraghavan=# set optimizer=on;                                                                                                                                                                                                                                                                                                                                             SET
> vraghavan=# explain select t2.organ_no from test t2 group by t2.organ_no;
> NOTICE:  bogus varno: 65001 (ruleutils.c:3228)
> ```

Oh, I can also reproduce it now. Don't know why I didn't see it earlier, 
must've been something funny in my environment.

- Heikki

**Comment by vraghavan78** [comment]
```
(lldb) bt
* thread #1: tid = 0x16e8ed, 0x000000010f9f39f6 postgres`get_variable(var=0x00007f947e89dd40, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 1094 at ruleutils.c:3233, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x000000010f9f39f6 postgres`get_variable(var=0x00007f947e89dd40, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 1094 at ruleutils.c:3233
    frame #1: 0x000000010f9f1140 postgres`get_rule_expr(node=0x00007f947e89dd40, context=0x00007fff508620b0, showimplicit='\x01') + 1216 at ruleutils.c:4036
    frame #2: 0x000000010f9f37f1 postgres`get_variable(var=0x00007f947e87daf0, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 577 at ruleutils.c:3189
    frame #3: 0x000000010f9f1140 postgres`get_rule_expr(node=0x00007f947e87daf0, context=0x00007fff508620b0, showimplicit='\x01') + 1216 at ruleutils.c:4036
    frame #4: 0x000000010f9f37f1 postgres`get_variable(var=0x00007f947e87d7e0, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 577 at ruleutils.c:3189
    frame #5: 0x000000010f9f1140 postgres`get_rule_expr(node=0x00007f947e87d7e0, context=0x00007fff508620b0, showimplicit='\x01') + 1216 at ruleutils.c:4036
    frame #6: 0x000000010f9f37f1 postgres`get_variable(var=0x00007f947e87d4d0, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 577 at ruleutils.c:3189
    frame #7: 0x000000010f9f1140 postgres`get_rule_expr(node=0x00007f947e87d4d0, context=0x00007fff508620b0, showimplicit='\x01') + 1216 at ruleutils.c:4036
    frame #8: 0x000000010f9f37f1 postgres`get_variable(var=0x00007f947e87d1c0, levelsup=0, istoplevel='\0', context=0x00007fff508620b0) + 577 at ruleutils.c:3189
    frame #9: 0x000000010f9f1140 postgres`get_rule_expr(node=0x00007f947e87d1c0, context=0x00007fff508620b0, showimplicit='\x01') + 1216 at ruleutils.c:4036
    frame #10: 0x000000010f9ef3db postgres`deparse_expression_pretty(expr=0x00007f947e87d1c0, dpcontext=0x00007f948008f738, forceprefix='\0', showimplicit='\x01', prettyFlags=1, startIndent=0) + 107 at ruleutils.c:1729
    frame #11: 0x000000010f9ef436 postgres`deparse_expr_sweet(expr=0x00007f947e87d1c0, dpcontext=0x00007f948008f738, forceprefix='\0', showimplicit='\x01') + 70 at ruleutils.c:1691
    frame #12: 0x000000010f5fcf7c postgres`show_grouping_keys(plan=0x00007f947e87cc60, numCols=1, subplanColIdx=0x00007f947e89e380, qlabel="Group By", str=0x00007fff50862c28, indent=3, es=0x00007f947e89f198) + 1004 at explain.c:2106
    frame #13: 0x000000010f5fb17a postgres`explain_outNode(str=0x00007fff50862c28, plan=0x00007f947e87cc60, planstate=0x00007f948008d900, outer_plan=0x0000000000000000, parentPlan=0x00007f947e87c950, indent=3, es=0x00007f947e89f198) + 5898 at explain.c:1563
    frame #14: 0x000000010f5fb8ef postgres`explain_outNode(str=0x00007fff50862c28, plan=0x00007f947e87c950, planstate=0x00007f948008cc90, outer_plan=0x0000000000000000, parentPlan=0x0000000000000000, indent=0, es=0x00007f947e89f198) + 7807 at explain.c:1763
    frame #15: 0x000000010f5f940f postgres`ExplainOnePlan(plannedstmt=0x00007f947e87c818, params=0x0000000000000000, stmt=0x00007f947f94c950, queryString="explain\nselect t2.organ_no\nfrom test t2\ngroup by t2.organ_no;", tstate=0x00007f947f04f928) + 2943 at explain.c:609
    frame #16: 0x000000010f5f8794 postgres`ExplainOneQuery(query=0x00007f947f04e0f8, stmt=0x00007f947f94c950, queryString="explain\nselect t2.organ_no\nfrom test t2\ngroup by t2.organ_no;", params=0x0000000000000000, tstate=0x00007f947f04f928) + 212 at explain.c:294
    frame #17: 0x000000010f5f85bc postgres`ExplainQuery(stmt=0x00007f947f94c950, queryString="explain\nselect t2.organ_no\nfrom test t2\ngroup by t2.organ_no;", params=0x0000000000000000, dest=0x00007f947f027250) + 236 at explain.c:170
    frame #18: 0x000000010f91029f postgres`ProcessUtility(parsetree=0x00007f947f94c950, queryString="explain\nselect t2.organ_no\nfrom test t2\ngroup by t2.organ_no;", params=0x0000000000000000, isTopLevel='\x01', dest=0x00007f947f027250, completionTag="") + 6975 at utility.c:1559
    frame #19: 0x000000010f90dcb7 postgres`PortalRunUtility(portal=0x00007f947e80d260, utilityStmt=0x00007f947f94c950, isTopLevel='\x01', dest=0x00007f947f027250, completionTag="") + 391 at pquery.c:1393
    frame #20: 0x000000010f90c9a2 postgres`FillPortalStore(portal=0x00007f947e80d260, isTopLevel='\x01') + 226 at pquery.c:1266
    frame #21: 0x000000010f90c137 postgres`PortalRun(portal=0x00007f947e80d260, count=9223372036854775807, isTopLevel='\x01', dest=0x00007f947f94cdf8, altdest=0x00007f947f94cdf8, completionTag="") + 983 at pquery.c:1007
    frame #22: 0x000000010f904f3e postgres`exec_simple_query(query_string="explain\nselect t2.organ_no\nfrom test t2\ngroup by t2.organ_no;", seqServerHost=0x0000000000000000, seqServerPort=-1) + 2254 at postgres.c:1808
    frame #23: 0x000000010f903248 postgres`PostgresMain(argc=1, argv=0x00007f947f809dd0, dbname="frahman", username="frahman") + 3496 at postgres.c:4980
    frame #24: 0x000000010f880723 postgres`BackendRun(port=0x00007f947e40b0e0) + 723 at postmaster.c:6685
    frame #25: 0x000000010f87d741 postgres`BackendStartup(port=0x00007f947e40b0e0) + 369 at postmaster.c:6371
    frame #26: 0x000000010f87963f postgres`ServerLoop + 1247 at postmaster.c:2437
    frame #27: 0x000000010f8779b5 postgres`PostmasterMain(argc=15, argv=0x00007f947e4091b0) + 4949 at postmaster.c:1523
    frame #28: 0x000000010f75da44 postgres`main(argc=15, argv=0x00007f947e4091b0) + 788 at main.c:206
    frame #29: 0x00007fffad338255 libdyld.dylib`start + 1
    frame #30: 0x00007fffad338255 libdyld.dylib`start + 1
(lldb) f 13
frame #13: 0x000000010f5fb17a postgres`explain_outNode(str=0x00007fff50862c28, plan=0x00007f947e87cc60, planstate=0x00007f948008d900, outer_plan=0x0000000000000000, parentPlan=0x00007f947e87c950, indent=3, es=0x00007f947e89f198) + 5898 at explain.c:1563
   1560             show_upper_qual(plan->qual,
   1561                             "Filter", plan,
   1562                             str, indent, es);
-> 1563              show_grouping_keys(plan,
   1564                                ((Agg *) plan)->numCols,
   1565                                ((Agg *) plan)->grpColIdx,
   1566                                "Group By",
(lldb) p *plan
(Plan) $0 = {
  type = T_Agg
  plan_node_id = 2
  plan_parent_node_id = 1
  startup_cost = 0
  total_cost = 431.50953399999997
  plan_rows = 1
  plan_width = 3
  targetlist = 0x00007f947e87cd98
  qual = 0x0000000000000000
  lefttree = 0x00007f947e87cf70
  righttree = 0x0000000000000000
  initPlan = 0x0000000000000000
  extParam = 0x0000000000000000
  allParam = 0x0000000000000000
  flow = 0x00007f947f987b50
  dispatch = DISPATCH_UNDETERMINED
  directDispatch = {
    isDirectDispatch = '\0'
    contentIds = 0x0000000000000000
  }
  nMotionNodes = 1
  nInitPlans = 0
  sliceTable = 0x0000000000000000
  operatorMemKB = 100
  memoryAccountId = 5
}
(lldb) p nodeToString(plan->targetlist)
(char *) $1 = 0x00007f948008fbc8 "({TARGETENTRY :expr {VAR :varno 65001 :varattno 1 :vartype 1043 :vartypmod -1 :varlevelsup 0 :varnoold 1 :varoattno 1} :resno 1 :resname organ_no :ressortgroupref 0 :resorigtbl 17035 :resorigcol 1 :resjunk false})"
(lldb)
```

Something to do with the Agg node.

**Comment by vraghavan78** [comment]
on 4.3.11.3 it works.

```
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
   ->  GroupAggregate  (cost=0.00..431.51 rows=1 width=3)
         Group By: organ_no
         ->  Sort  (cost=0.00..431.51 rows=1 width=3)
               Sort Key: organ_no
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
                     Hash Key: organ_no
                     ->  HashAggregate  (cost=0.00..431.51 rows=1 width=3)
                           Group By: organ_no
                           ->  Sequence  (cost=0.00..431.08 rows=3334 width=3)
                                 ->  Partition Selector for test (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
                                       Partitions selected:  2 (out of 2)
                                 ->  Dynamic Table Scan on test (dynamic scan id: 1)  (cost=0.00..431.08 rows=3334 width=3)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.5.0
```

**Comment by vraghavan78** [comment]
On GPDB 5, I removed the Assert and this is the explain I get.

```
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
   ->  GroupAggregate  (cost=0.00..431.51 rows=1 width=3)
         Group By: organ_no
         ->  Sort  (cost=0.00..431.51 rows=1 width=3)
               Sort Key: organ_no
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
                     Hash Key: organ_no
                     ->  HashAggregate  (cost=0.00..431.51 rows=1 width=3)
                           Group By:
                           ->  Sequence  (cost=0.00..431.08 rows=3334 width=3)
                                 ->  Partition Selector for test (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
                                       Partitions selected: 2 (out of 2)
                                 ->  Dynamic Table Scan on test (dynamic scan id: 1)  (cost=0.00..431.08 rows=3334 width=3)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.5.0
```

The problem is how explain show the local aggregate.

```
                     ->  HashAggregate  (cost=0.00..431.51 rows=1 width=3)
                           Group By:
```

It is unable to print the grouping column. The following is the local agg node

```
:lefttree
                   {AGG
                   :plan_node_id 5
                   :plan_parent_node_id 4
                   :startup_cost 0.00
                   :total_cost 431.51
                   :plan_rows 1
                   :plan_width 3
                   :targetlist (
                      {TARGETENTRY
                      :expr
                         {VAR
                         :varno 65001
                         :varattno 1
                         :vartype 1043
                         :vartypmod -1
                         :varlevelsup 0
                         :varnoold 1
                         :varoattno 1
                         }
                      :resno 1
                      :resname organ_no
                      :ressortgroupref 0
                      :resorigtbl 24577
                      :resorigcol 1
                      :resjunk false
                      }
                   )
                   :qual <>
                   :extParam (b)
                   :allParam (b)
                   :flow
                      {FLOW
                      :flotype 0
                      :req_move 0
                      :locustype 0
                      :segindex 0
                      :numSortCols 0
                      :numOrderbyCols 0
                      :hashExpr <>
                      :flow_before_req_move <>
                      }
                   :dispatch 0
                   :nMotionNodes 0
                   :nInitPlans 0
                   :sliceTable <>
                   :lefttree
                      {SEQUENCE
                      :plan_node_id 6
                      :plan_parent_node_id 5
                      :startup_cost 0.00
                      :total_cost 431.08
                      :plan_rows 10000
                      :plan_width 3
                      :targetlist (
                         {TARGETENTRY
                         :expr
                            {VAR
                            :varno 65001
                            :varattno 1
                            :vartype 1043
                            :vartypmod -1
                            :varlevelsup 0
                            :varnoold 1
                            :varoattno 1
                            }
                         :resno 1
                         :resname organ_no
                         :ressortgroupref 0
                         :resorigtbl 24577
                         :resorigcol 1
                         :resjunk false
                         }
                      )
                      :qual <>
                      :extParam (b)
                      :allParam (b)
                      :flow <>
                      :dispatch 0
                      :nMotionNodes 0
                      :nInitPlans 0
                      :sliceTable <>
                      :lefttree <>
                      :righttree <>
                      :initPlan <>
                      :operatorMemKB 100
                      :subplans (
                         {PartitionSelector
                         :relid 24577
                         :nLevels 1
                         :scanId 1
                         :selectorId 0
                         :levelEqExpressions (<>)
                         :levelExpressions (<>)
                         :residualPredicate <>
                         :propagationExpression
                            {CONST
                            :consttype 23
                            :consttypmod -1
                            :constlen 4
                            :constbyval true
                            :constisnull false
                            :constvalue 4 [ 1 0 0 0 0 0 0 0 ]
                            }
                         :printablePredicate <>
                         :staticSelection true
                         :staticPartOids (o 24584 24606)
                         :staticScanIds (i 1 1)
                         :plan_node_id 8
                         :plan_parent_node_id 6
                         :startup_cost 10.00
                         :total_cost 100.00
                         :plan_rows 100
                         :plan_width 4
                         :targetlist (
                            {TARGETENTRY
                            :expr
                               {PARTOIDEXPR
                               :level 0
                               }
                            :resno 1
                            :resname ColRef_0009
                            :ressortgroupref 0
                            :resorigtbl 0
                            :resorigcol 0
                            :resjunk false
                            }
                         )
                         :qual <>
                         :extParam (b)
                         :allParam (b)
                         :flow <>
                         :dispatch 0
                         :nMotionNodes 0
                         :nInitPlans 0
                         :sliceTable <>
                         :lefttree <>
                         :righttree <>
                         :initPlan <>
                         :operatorMemKB 100
                         }
                         {DYNAMICTABLESCAN
                         :plan_node_id 7
                         :plan_parent_node_id 6
                         :startup_cost 0.00
                         :total_cost 431.08
                         :plan_rows 10000
                         :plan_width 3
                         :targetlist (
                            {TARGETENTRY
                            :expr
                               {VAR
                               :varno 1
                               :varattno 1
                               :vartype 1043
                               :vartypmod -1
                               :varlevelsup 0
                               :varnoold 1
                               :varoattno 1
                               }
                            :resno 1
                            :resname organ_no
                            :ressortgroupref 0
                            :resorigtbl 24577
                            :resorigcol 1
                            :resjunk false
                            }
                         )
                         :qual <>
                         :extParam (b)
                         :allParam (b)
                         :flow <>
                         :dispatch 0
                         :nMotionNodes 0
                         :nInitPlans 0
                         :sliceTable <>
                         :lefttree <>
                         :righttree <>
                         :initPlan <>
                         :operatorMemKB 100
                         :scanrelid 1
                         :partIndex 1
                         :partIndexPrintable 1
                         :partIndex 1
                         :partIndexPrintable 1
                         }
                      )
                      }
                   :righttree <>
                   :initPlan <>
                   :operatorMemKB 63800
                   :aggstrategy 2
                   :numCols 1
                   :grpColIdx 1
                   :grpOperators 98
                   :numGroups 0
                   :transSpace 0
                   :numNullCols 0
                   :inputGrouping 0
                   :grouping 0
                   :inputHasGrouping false
                   :rollupGSTimes 0
                   :lastAgg false
                   :streaming true
                   }
```

@hlinnaka I do not see anything incorrect.

**Comment by hlinnaka** [comment]
On 02/08/2017 10:27 PM, Venkatesh Raghavan wrote:
> On GPDB 5, I removed the Assert and this is the explain I get.
>
> ```
>  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
>    ->  GroupAggregate  (cost=0.00..431.51 rows=1 width=3)
>          Group By: organ_no
>          ->  Sort  (cost=0.00..431.51 rows=1 width=3)
>                Sort Key: organ_no
>                ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.51 rows=1 width=3)
>                      Hash Key: organ_no
>                      ->  HashAggregate  (cost=0.00..431.51 rows=1 width=3)
>                            Group By:
>                            ->  Sequence  (cost=0.00..431.08 rows=3334 width=3)
>                                  ->  Partition Selector for test (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
>                                        Partitions selected: 2 (out of 2)
>                                  ->  Dynamic Table Scan on test (dynamic scan id: 1)  (cost=0.00..431.08 rows=3334 width=3)
>  Settings:  optimizer=on
>  Optimizer status: PQO version 2.5.0
> ```
>
> The problem is how explain show the local aggregate.
>
> ```
>                      ->  HashAggregate  (cost=0.00..431.51 rows=1 width=3)
>                            Group By:
> ```
>
> It is unable to print the grouping column. The following is the local agg node

The plan with optimizer=off is almost identical. I think the culprit 
here is having a Sequence node, immediately below the Agg node. Sequence 
nodes are a bit special. If a Sequence node has a Partition Selector as 
the first child, like above, we shouldn't resolve the Var down to the 
Partition Selector node, but into the second (or last?) child. The 
push_plan() function in ruleutils.c has some magic to deal with that.

Looking at show_grouping_keys(), there is code similar to that in 
push_plan(), to deal with Append nodes. But it doesn't do anything 
special about Sequence nodes. However, it's not clear to me why 
show_grouping_keys() does anything special about Append nodes either. 
For comparison show_sort_keys() doesn't do that, and it works fine.

I tried removing the special handling of Append nodes in 
show_grouping_keys(), and made it call deparse_context_for_plan() with 
the Agg nodes' immediate child node, rather than the child's chidlren. 
In other words, I made it work like show_sort_keys() does. This fixed 
the reported test case, but there are now some other "bogus varno" 
failures in the regression suite.

I'm a bit baffled on why special code was put in show_grouping_keys() in 
the first place. I'd like to fix this by removing it, but will need to 
investigate those other failures...

- Heikki

**Comment by hlinnaka** [comment]
This code will be superseded by similar code in PostgreSQL 9.4. Given that, I'm not going spend more time figuring out why show_grouping_keys() works differently from show_sort_keys(). I think by the time we get to 9.4, we'll know better.

For now, let's just add special handling of Sequence nodes into show_grouping_keys(), similar to  push_plan(). PR #1713 implements that.

**Comment by hlinnaka** [comment]
Pushed PR #1713 to fix this. Thanks for the report @vincexiaoqiang!
